### ISSUE TITLE
Header template-part disappears from template after theme change

### ISSUE THREAD

**Issue by kmanijak** [woothemes/woocommerce#42181]
### Description

The header template part disappears on the frontend after updating the theme, when used in other templates.

### Steps to reproduce

1. On TT3
2. Go to Product Catalog template
3. Make sure you have Header template-part included in the template
4. Make ANY change to Header template-part and save it so it's customised
5. Update to TT4
6. Go to frontend (`/shop`)

Expected: Header is still there
Actual: Header is missing

Editor | FE
-- | --
<img width="938" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/e0af62b9-6848-4794-8e7d-d6f1aa9961a1"> |  <img width="1204" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/214bd46d-d855-4543-a068-918a6f4e4c77">
 

### More details

When entering the code editor you can see the template part has a theme assigned:

`<!-- wp:template-part {"slug":"header","theme":"twentytwentythree"} /-->`

After updating the theme value manually in the code editor to `twentytwentyfour` it appears on the frontend

**Comment by danieldudzic** [comment]
I have tested this, but I'm unable to replicate the issue.

The _new_ theme correctly loads its own `Header` template and the old template no longer is relevant (on my test site).

**Comment by danieldudzic** [comment]
@sunyatasattva Are you able to replicate this? I'd appreciate it if you could have a look. 🙏

**Comment by sunyatasattva** [comment]
Hey @danieldudzic and @kmanijak, I think I see what's the problem here. It seems like when you activate a theme you have a template part that belongs to that theme. If you then modify a template containing that template part, and *then* switch theme, the problem appears: probably because the theme doesn't want to override any user changes.

This is probably a bug in core, to be honest.

**Comment by danieldudzic** [comment]
I don't think this is a bug. As you mentioned, a theme-related template part belongs to that theme. If the new theme provides its own template part, it should take priority over the customized one (from the previous theme).

**Comment by sunyatasattva** [comment]
Mm, I think we are talking about two different things right now, perhaps I didn't express myself correctly. You say:

> If the new theme provides its own template part, it should take priority over the customized one (from the previous theme).

That's exactly what is not happening. The new theme template part **won't** take priority over a customized one.

Also, I'm not sure I agree that it should, and this is a much more nuanced conversation than what's happening here, albeit very similar: what happens here is that if someone customized a template which _includes_ a template part, and then changes a theme, that template part won't be updated, causing an error, as it's not available anymore.

This is why I think this is a bug. I hope I expressed myself correctly here.

**Comment by danieldudzic** [comment]
I see what you are saying. Thank you for clarifying the situation for me. I was now able to replicate the issue.

I have tried replicating this with WooCommerce disabled but wasn't able to.

I was able to replicate it with 6.3.2 and TT3.

Do you have any idea where the bug could be originating from?

**Comment by kmanijak** [comment]
Thanks for adding more info! I added the Header customization as a necessary step in steps to reproduce.

> I have tried replicating this with WooCommerce disabled but wasn't able to.

@danieldudzic, in such a case my guess would be `BlockTemplatesController`!

**Comment by Aljullu** [comment]
For me, in order to reproduce, I had to:

1. On TT3
2. Go to Product Catalog template
3. Make sure you have Header template-part included in the template
4. Make ANY change to Header template-part and save it so it's customised.
5. **Make also a change to the Product Catalog template and save it so it's customised.**
6. Update to TT4
7. Go to frontend (`/shop`)

Not sure if it's only me, but without step 5 I couldn't reproduce. :sweat: 

---

I also noticed a couple of things:

1. The issue can be reproduced with other themes. So I would say it's not specific to Twenty-Twenty Four.
2. The issue can't be reproduced with non-WooCommerce templates. That's because when modifying a non-WC template, those modifications are saved "attached" to the theme, so when switching to another theme, the default template is loaded instead of the modified one.

IMO the best solution would be to do the same with WooCommerce templates: so when modifying a WooCommerce template, it's saved "attached" to the theme and, when switching themes, the default template is loaded again. Switching back to the initial theme, should load the modified template. Does that sound like a good approach?

cc @gigitux @tjcafferkey as this seems related to the project to refactor of BlockTemplatesController.

**Comment by gigitux** [comment]
@Aljullu Yes, this issue will be addressed with the project about BlockTemplatesController. @nielslange opened a similar issue https://github.com/woocommerce/woocommerce-blocks/issues/11106.

I want to highlight that this is not a regression introduced by us in the last weeks or for some changes in WordPress 6.4.

I think that we have had this wrong behavior for a long time, so even if we **must** fix it, should we keep the label `high priority`?

**Comment by danieldudzic** [comment]
I'm comfortable with removing the `high priority` label. This issue should probably also be moved from the `6.4 & TT4` epic to the `BlockTemplatesController` project.

**Comment by gigitux** [comment]
> I'm confortable with removing the `high priority` label. This issue should probably also be moved from the `6.4 & TT4` epic to the `BlockTemplatesController` project.

It makes sense! The issue has been added to woocommerce/woocommerce#42251! Thanks, @tjcafferkey!

**Comment by Aljullu** [comment]
Heads-up that I unassigned myself from this issue, as I'm not actively working on it. There was a work-around here: https://github.com/woocommerce/woocommerce/pull/46831. Otherwise this issue will be fixed "for free" once we migrate to the template registration API introduced in Gutenberg (https://github.com/woocommerce/woocommerce/issues/47885), because customized templates will be attached to the theme.

**Comment by gigitux** [comment]
Cool! Thanks for the heads-up and for your work! I would keep this issue open, or do you think that we should close it?

**Comment by Aljullu** [comment]
The issue is still reproducible, so I think we should keep it open until it's fixed, either via a work-around or with https://github.com/woocommerce/woocommerce/issues/47885.

**Comment by mikejolley** [comment]
I am today running into this issue testing TT5 theme. It only affects some product templates.

![Image](https://github.com/user-attachments/assets/baa68cac-5f05-42cc-b098-aa7325c18db3)

Unfortunately there is no way to reset the product pages in question to default, so the only way to fix is to remove this template part and reinsert the header template.
